src/factories/calorimetry/CalorimeterClusterRecoCoG_factory.h

```diff
@@ -24,19 +27,41 @@ class CalorimeterClusterRecoCoG_factory : public JOmniFactory<CalorimeterCluster
     PodioOutput<edm4eic::Cluster> m_cluster_output {this};
     PodioOutput<edm4eic::MCRecoClusterParticleAssociation> m_assoc_output {this};
 
-    ParameterRef<std::string> m_energyWeight {this, "energyWeight", config().energyWeight};
-    ParameterRef<double> m_samplingFraction {this, "samplingFraction", config().sampFrac};
-    ParameterRef<double> m_logWeightBase {this, "logWeightBase", config().logWeightBase};
-    ParameterRef<bool> m_enableEtaBounds {this, "enableEtaBounds", config().enableEtaBounds};
+    std::vector<std::unique_ptr<ParameterBase>> m_parameters;
 
     Service<DD4hep_service> m_geoSvc {this};
     // Resource<DD4hep_service, const dd4hep::Detector*> m_detector {this, [](std::shared_ptr<DD4hep_service> s, int64_t run_nr){ return s->detector(); }}
 
 public:
     void Configure() {
-        m_algo = std::make_unique<AlgoT>(GetPrefix());
-        m_algo->applyConfig(config());
-        m_algo->init(m_geoSvc().detector(), logger());
+      m_algo = std::make_unique<AlgoT>(GetPrefix());
+
+      for (const auto& [key, prop] : m_algo->getProperties()) {
```

Comment:
We need m_algo to exist before we can register the upstream parameters. We need to have ownership of the parameters in this factory since JOmniFactory does not clear the pointers in its m_parameters. We can probably do something like RegisterAlgorithmParameters(m_algo, m_parameters), but we'd still need to declare the vector<unique_ptr> as part of this algorithm. A mixin class may be another option.

Reply:
Moved this into JOmniFactory in [0211205](https://github.com/eic/EICrecon/pull/1165/commits/0211205b1a5caff56968fe543e503ca6686b850b)